bundles/org.openhab.binding.systeminfo/src/main/java/org/openhab/binding/systeminfo/internal/handler/SysteminfoHandler.java

```diff
@@ -91,31 +103,59 @@ public class SysteminfoHandler extends BaseThingHandler {
      */
     public static final int WAIT_TIME_CHANNEL_ITEM_LINK_INIT = 1;
 
+    /**
+     * String used to extend thingUID and channelGroupTypeUID for thing definition with added dynamic channels and
+     * extended channels. It is set in the constructor and unique to the thing.
+     */
+    public final String idExtString;
+
+    public final SysteminfoThingTypeProvider thingTypeProvider;
+
     private SysteminfoInterface systeminfo;
 
     private @Nullable ScheduledFuture<?> highPriorityTasks;
     private @Nullable ScheduledFuture<?> mediumPriorityTasks;
 
-    private Logger logger = LoggerFactory.getLogger(SysteminfoHandler.class);
+    /**
+     * Caches for cpu process load and process load for a given pid. Using this cache limits the process load refresh
+     * interval to the minimum interval. Too frequent refreshes leads to inaccurate results. This could happen when the
+     * same process is tracked as current process and as a channel with pid parameter, or when the task interval is set
+     * too low.
+     */
+    private static final int MIN_PROCESS_LOAD_REFRESH_INTERVAL_MS = 2000;
+    private ExpiringCache<PercentType> cpuLoadCache = new ExpiringCache<>(MIN_PROCESS_LOAD_REFRESH_INTERVAL_MS,
+            () -> getSystemCpuLoad());
+    private ExpiringCacheMap<Integer, @Nullable DecimalType> processLoadCache = new ExpiringCacheMap<>(
+            MIN_PROCESS_LOAD_REFRESH_INTERVAL_MS);
 
-    public SysteminfoHandler(Thing thing, @Nullable SysteminfoInterface systeminfo) {
+    private final Logger logger = LoggerFactory.getLogger(SysteminfoHandler.class);
+
+    public SysteminfoHandler(Thing thing, SysteminfoThingTypeProvider thingTypeProvider,
+            SysteminfoInterface systeminfo) {
         super(thing);
-        if (systeminfo != null) {
-            this.systeminfo = systeminfo;
-        } else {
-            throw new IllegalArgumentException("No systeminfo service was provided");
-        }
+        this.thingTypeProvider = thingTypeProvider;
+        this.systeminfo = systeminfo;
+
+        idExtString = "-" + thing.getUID().getId();
     }
 
     @Override
     public void initialize() {
+        logger.trace("Initializing thing {} with thing type {}", thing.getUID().getId(),
+                thing.getThingTypeUID().getId());
+        restoreChannelsConfig(); // After a thing type change, previous channel configs will have been stored, and will
+                                 // be restored here.
         if (instantiateSysteminfoLibrary() && isConfigurationValid() && updateProperties()) {
-            groupChannelsByPriority();
-            scheduleUpdates();
-            updateStatus(ThingStatus.ONLINE);
+            if (!addDynamicChannels()) { // If there are new channel groups, the thing will get recreated with a new
```

Comment:
Indeed, I missed that. I don't think it can happen, as the computer thing type is the default one from the xml definition, but I added a thing status update to cope with it if it would.